fivox/uriHandler.cpp

```diff
@@ -432,54 +423,99 @@ size_t URIHandler::getSizeInVoxel() const
     return _impl->getSizeInVoxel();
 }
 
+bool URIHandler::showProgress() const
+{
+    return _impl->showProgress();
+}
+
 template< class T > itk::SmartPointer< ImageSource< itk::Image< T, 3 >>>
 URIHandler::newImageSource() const
 {
-    LBINFO << "Loading events..." << std::endl;
-
-    itk::SmartPointer< ImageSource< itk::Image< T, 3 >>> source =
-        ImageSource< itk::Image< T, 3 >>::New();
-    std::shared_ptr< EventFunctor< itk::Image< T, 3 >>> functor =
-        _newFunctor< T >( *this );
-
-    EventSourcePtr loader = _newLoader( *this );
+    EventSourcePtr eventSource = newEventSource();
 
-    LBINFO << loader->getNumEvents() << " events " << *this << ", dt = "
-           << loader->getDt() << " ready to voxelize" << std::endl;
-
-    if( _impl->showProgress( ))
-        source->showProgress();
+    itk::SmartPointer< ImageSource< itk::Image< T, 3 >>> source;
+    if( getFunctorType() == FunctorType::density )
+        source = DensityImageSource< itk::Image< T, 3 >>::New();
+    else
+    {
+        auto functorSource = FunctorImageSource< itk::Image< T, 3 >>::New();
+        auto functor = newFunctor< T >();
+        functorSource->setFunctor( functor );
+        functor->setEventSource( eventSource );
+        source = functorSource;
+    }
 
-    functor->setSource( loader );
-    source->setFunctor( functor );
+    source->setEventSource( eventSource );
     source->setup( *this );
     return source;
 }
 
+EventSourcePtr URIHandler::newEventSource() const
+{
+    switch( getType( ))
+    {
+    case VolumeType::compartments:
+        return std::make_shared< CompartmentLoader >( *this );
+    case VolumeType::somas:
+        return std::make_shared< SomaLoader >( *this );
+    case VolumeType::spikes:
+        return std::make_shared< SpikeLoader >( *this );
+    case VolumeType::synapses:
+        return std::make_shared< SynapseLoader >( *this );
+    case VolumeType::test:
+        return std::make_shared< TestLoader >( *this );
+    case VolumeType::vsd:
+        return std::make_shared< VSDLoader >( *this );
+    default:
+        return nullptr;
+    }
+}
+
+template< class T > std::shared_ptr< EventFunctor< itk::Image< T, 3 >>>
+URIHandler::newFunctor() const
+{
+    switch( getFunctorType( ))
+    {
+    case FunctorType::density:
+        return std::make_shared< DensityFunctor< itk::Image< T, 3 >>>();
```

Comment:
What was the advantage of make_shared? That the implementation can allocate the memory for the object and the count in a single chunk?